package/libskylark/src/libskylark.c

```diff
@@ -0,0 +1,333 @@
+{
+  sbp_zmq_state_t *sbp_zmq_state = (sbp_zmq_state_t *)context;
+  sbp_zmq_message_send(sbp_zmq_state, SBP_MSG_BASE_POS_LLH, len, msg);
+}
+
+static void base_ecef_callback(u16 sender_id, u8 len, u8 msg[], void *context)
+{
+  sbp_zmq_state_t *sbp_zmq_state = (sbp_zmq_state_t *)context;
+  sbp_zmq_message_send(sbp_zmq_state, SBP_MSG_BASE_POS_ECEF, len, msg);
+}
+
+u32 msg_read(u8 *buf, u32 n, void *context)
+{
+  int *fd = (int *)context;
+  ssize_t m = read(*fd, buf, n);
+  return m;
+}
+
+RC download_io_loop(skylark_client_config_t *config)
+{
+  // TODO sender id is not going to be right here, since we're relaying
+  // messages.
+  sbp_zmq_state_t sbp_zmq_state;
+  if (sbp_zmq_init(&sbp_zmq_state, config->sbp_zmq_config) < 0) {
+    exit(EXIT_FAILURE);
+  }
+  sbp_state_t sbp_state;
+  sbp_msg_callbacks_node_t callback_node;
+  sbp_state_init(&sbp_state);
+  sbp_state_set_io_context(&sbp_state, &config->fd_in);
+  sbp_register_callback(&sbp_state, SBP_MSG_OBS,
+                        &base_obs_callback, &sbp_zmq_state,
+                        &callback_node);
+  sbp_msg_callbacks_node_t base_ecef_callback_node;
+  sbp_register_callback(&sbp_state, SBP_MSG_BASE_POS_ECEF,
+                        &base_ecef_callback, &sbp_zmq_state,
+                        &base_ecef_callback_node);
+  sbp_msg_callbacks_node_t base_llh_callback_node;
+  sbp_register_callback(&sbp_state, SBP_MSG_BASE_POS_LLH,
+                        &base_llh_callback, &sbp_zmq_state,
+                        &base_llh_callback_node);
+  for (;;) {
+    sbp_process(&sbp_state, &msg_read);
+  }
+  sbp_zmq_deinit(&sbp_zmq_state);
+  return NO_ERROR;
+}
+
+/**
+ *  Upload processes, for base stations and reference station processing.
+ */
+
+static size_t upload_callback(void *p, size_t size, size_t n, void *up)
+{
+  int *fd = (int *)up;
+  ssize_t m = read(*fd, p, size * n);
+  if (m < 0) {
+    return CURL_READFUNC_ABORT;
+  }
+  return m;
+}
+
+RC upload_process(skylark_client_config_t *config)
+{
+  CURLcode res = curl_global_init(CURL_GLOBAL_ALL);
+  if (res != CURLE_OK) {
+    exit(EXIT_FAILURE);
+  }
+  CURL *curl = curl_easy_init();
+  if (curl == NULL) {
+    curl_global_cleanup();
+    exit(EXIT_FAILURE);
+  }
+  curl_easy_setopt(curl, CURLOPT_URL, config->endpoint_url);
+  curl_easy_setopt(curl, CURLOPT_PUT, 1L);
+  curl_easy_setopt(curl, CURLOPT_READFUNCTION, &upload_callback);
+  curl_easy_setopt(curl, CURLOPT_READDATA, &config->fd_in);
+#ifdef VERBOSE
+  curl_easy_setopt(curl, CURLOPT_VERBOSE, 1L);
+#endif
+  curl_easy_setopt(curl, CURLOPT_USERAGENT, USER_AGENT);
+  struct curl_slist *chunk = NULL;
+  chunk = curl_slist_append(chunk, config->encoding);
+  chunk = curl_slist_append(chunk, config->content_type_header);
+  chunk = curl_slist_append(chunk, config->device_header);
+  curl_easy_setopt(curl, CURLOPT_HTTPHEADER, chunk);
+  res = curl_easy_perform(curl);
+  if (res != CURLE_OK) {
```

Comment:
It seems to make sense to have some recovery here - and not kicking out of the process with every failure. There should be a limit of retries, in case something's gone wrong with the process - but if there's glitches in the network showing here, seems heavyweight to let monit handle it.